docs/development/zod.md

```diff
@@ -0,0 +1,232 @@
+Use helpers from `schema-utils.ts` for this purpose.
+
+Here is an **incorrect** way to parse from string:
+
+```ts
+try {
+  const json = JSON.parse(input);
+  return ApiResult.parse(json);
+} catch (e) {
+  return null;
+}
+```
+
+The **correct** way to parse from string:
+
+```ts
+Json.pipe(ApiResult).parse(input);
+```
+
+### Use `.transform()` method to process parsed data
+
+Schema validation helps to be more confident with types during downstream data transformation.
+
+You can go even further and perform some transformations as the part of schema itself.
+
+This is an example of **undesired** data transformation:
+
+```ts
+const Box = z.object({
+  width: z.number(),
+  height: z.number(),
+  length: z.number(),
+});
+
+const { width, height, length } = Box.parse(input);
+const volume = width * height * length;
+```
+
+Instead, use more idiomatic `.tranform()` method:
+
+```ts
+const Volume = z
+  .object({
+    width: z.number(),
+    height: z.number(),
+    length: z.number(),
+  })
+  .transform(({ width, height, length }) => width * height * length);
+
+Volume.parse({
+  width: 10,
+  height: 20,
+  length: 125,
+});
+```
+
+### Separate validation from transformation
+
+When parsing third party data, we are typically doing the following:
+
+- Validating that the data is correct/sufficient to use
+- Transforming it into a standardized format for internal usage
+
+Although it's not a strict requirement, your code will be cleaner if you perform the validation step first and then follow with transformation.
+
+### Stick to permissive behavior when possible
+
+Zod schemas are strict, and even some insufficient field is incorrect, the whole data will be treated like malformed.
+This could lead to cases when Renovate could've continued processing, but didn't.
+
+Remember: our goal is not to validate that data corresponds to any official specifications, but rather to ensure that the data is enough for Renovate to use.
+
+There are some techniques to make it more permissive to the input data.
+
+#### Use `.catch()` to force default values
+
+```ts
+const Box = z.object({
+  width: z.number().catch(10),
+  height: z.number().catch(10),
+});
+
+const box = Box.parse({ width: 10, height: null });
+// => { width: 10, height: 10 }
+```
+
+#### Use `LooseArray` and `LooseRecord` to filter out incorrect values from collections
+
+Suppose you want to validate an array and retain only number values in it.
+By using only methods provided by `zod` library, you'll have to write something like this:
+
+```ts
+const OnlyNumbers = z
+  .array(z.union([z.number(), z.null()]).catch(null))
+  .transform((xs) => xs.filter((x): x is number => x !== null));
+```
+
+While the problem is common, the code is quite complicated.
+
+Instead, you should use `LooseArray` and `LooseRecord` helpers provided in `schema-utils.ts`:
+
+```ts
+const OnlyNumbers = LooseArray(z.number());
+```
+
+[TODO: more details on how the above would be used]
```

Comment:
OnlyNumbers would need to be used somewhere else to be useful, right?

Reply:
This is just an artificial example, I used it instead of something real but more verbose, like `ReleaseResult` filtering or something

Reply:
I found the example to be unhelpful because it was too artificial. Can you create a more realistic one?